content/shmem_test_all.tex

```diff
@@ -0,0 +1,63 @@
+\apisummary{
+    Indicate whether all variables within an array of variables on the local \ac{PE} meet a specified test condition.
+}
+
+\begin{apidefinition}
+
+\begin{C11synopsis}
+int @\FuncDecl{shmem\_test\_all}@(TYPE *ivars, size_t nelems, const int *status, int cmp, TYPE cmp_value);
+\end{C11synopsis}
+where \TYPE{} is one of the point-to-point synchronization types specified by
+Table \ref{p2psynctypes}.
+
+\begin{Csynopsis}
+int @\FuncDecl{shmem\_\FuncParam{TYPENAME}\_test\_all}@(TYPE *ivars, size_t nelems, const int *status, int cmp,
+    TYPE cmp_value);
+\end{Csynopsis}
+where \TYPE{} is one of the point-to-point synchronization types and has a
+corresponding \TYPENAME{} specified by Table \ref{p2psynctypes}.
+
+\begin{apiarguments}
+
+  \apiargument{IN}{ivars}{A pointer to an array of remotely accessible data
+    objects.}
+  \apiargument{IN}{nelems}{The number of elements in the \VAR{ivars} array.}
+  \apiargument{IN}{status}{An optional mask array of length \VAR{nelems}
+    that indicates which elements in \VAR{ivars} are excluded from the test set.}
+  \apiargument{IN}{cmp}{A comparison operator from Table~\ref{p2p-consts}
+    that compares elements of \VAR{ivars} with \VAR{cmp\_value}.}
+  \apiargument{IN}{cmp\_value}{The value to be compared with the objects
```

Comment:
I can see a variant of this API where we can provide a cmp_value per element as being useful.

That is, cmp_value would be a TYPE* with one value per element in ivars.

This would up the memory consumption, but in exchange for more flexibility.

I'm assuming you would agree that if we wanted to add that, it should be included in a separate follow on proposal?

Reply:
Yes - this has come up before, and it may have also been under the guise of the `cmp` operator as well as `cmp_value`.  But I imagine more relevant use-cases for `cmp_value`; for instance, in a work-sharing app where each `ivars` element has a corresponding (and varying) number of expected tasks.

In the past the consensus was to avoid the memory cost until we're certain that it's a useful feature.  Perhaps there's a clean way to make it optional, but I imagine its at least another argument to an already long list of arguments. 

I'd be happy to add this in if that's the actual consensus ([how about another Doodle poll?](https://doodle.com/poll/vfqvxfn59tnkz4hn)).  And yes, a follow on proposal is preferred if that's acceptable.

Reply:
I don't understand the poll. There's no option to keep the API as currently proposed (i.e., `cmp` is `int` and `cmp_value` is `TYPE`); the "other" option is "it depends / don't care". I think such a change should be should be in another proposal.

I think making `cmp` and `cmp_value` both `nelems`-long arrays would add far too much memory pressure for this routine to be effective in the cases I have in mind. (It's seriously making me think of struct-of-arrays vs. array-of-structs layouts and the cache behavior implied here.)

Reply:
Ah, thanks @nspark.  I tried to clarify the poll with these options for which direction to take the API:

1) Only `cmp_value` is an array (of length `nelems`)
2) Only `cmp` is an array
3) Both `cmp_value` and `cmp` are arrays
4) Neither `cmp_value` and `cmp` are arrays (_as currently proposed_, length of 1 each)
5) Explore other options, like making 1 and/or 2 optional.

And someone please let me know if I'm a little too poll-happy ;)

Reply:
And yes, the poll is intended to guide a (possible) separate proposal.